packages/cli/src/telemetry.ts

```diff
@@ -0,0 +1,146 @@
+    if (network && network.match(/dev-/)) network = 'development';
+
+    // encrypt data before sending it
+    const concealedData = concealData(commandName, options, telemetry.salt);
+    const commandData = network ? { ...concealedData, network } : concealedData;
+
+    await this.sendToFirebase(telemetry.uuid, commandData);
+  },
+
+  async sendToFirebase(uuid: string, commandData: any): Promise<void> {
+    // Initialize Firebase and anonymously authenticate
+    const app = firebase.initializeApp(FIREBASE_CONFIG);
+    const db = app.firestore();
+    const { FieldValue } = firebase.firestore;
+
+    try {
+      await app.auth().signInAnonymously();
+
+      // create a new command document for the current uuid
+      await db.runTransaction(async tx => {
+        const dbSnapshot = await tx.get(db.doc(`users/${uuid}`));
+        let incrementalId;
+        // if the current user document exists, retreive the latest command id and create a new command document.
+        // otherwise, create a document for the user and set the id to 0.
+        if (dbSnapshot.exists) {
+          incrementalId = dbSnapshot.get('latestId') + 1;
+          await tx.update(db.doc(`users/${uuid}`), { latestId: FieldValue.increment(1) });
+        } else {
+          incrementalId = 0;
+          await tx.set(db.doc(`users/${uuid}`), { latestId: 0 });
+        }
+        await tx.set(db.collection(`users/${uuid}/commands`).doc(), { ...commandData, id: incrementalId });
+      });
+
+      // close all connections
+      await app.delete();
+    } catch (_) {
+      return;
+    }
+  },
+};
+
+async function checkOptIn(interactive: boolean): Promise<GlobalTelemetryOptions | undefined> {
+  const project = new ProjectFile();
+  const localOptIn = project.telemetryOptIn;
+
+  const { data: globalDataDir } = envPaths('openzeppelin-sdk');
+  const globalDataPath = path.join(globalDataDir, 'telemetry.json');
+  let globalOptIn: GlobalTelemetryOptions | undefined = await fs.readJson(globalDataPath).catch(() => undefined);
+
+  if (localOptIn === false) return undefined;
+
+  // disable interactivity manually for tests and CI
+  if (DISABLE_INTERACTIVITY) interactive = false;
+
+  if (globalOptIn === undefined && interactive) {
+    const { telemetry } = await inquirer.prompt({
+      name: 'telemetry',
+      type: 'confirm',
+      message: 'telemetry?',
+      default: true,
+    });
+
+    const salt = crypto.randomBytes(32);
+    globalOptIn = { optIn: telemetry, uuid: uuid(), salt: salt.toString('hex') };
+    await fs.ensureDir(globalDataDir);
+    await fs.writeJson(globalDataPath, globalOptIn);
+  }
+
+  if (localOptIn === undefined && globalOptIn !== undefined) {
+    project.telemetryOptIn = globalOptIn.optIn;
+    // following function is sync
+    project.write();
+  }
+
+  return globalOptIn;
+}
+
+function hashField(field: Field, salt: string): string {
+  const hash = crypto.createHash('sha256');
+  hash.update(salt);
```

Comment:
What kind of scenario would we be trying to cover for here?

I'm not sure it's worth checking for the empty string. It would be just as bad if we were using any other non-random salt, and there's no way to check for all of them.